### PR TITLE
feat: benchmark foreign investments worst case

### PR DESCRIPTION
**Pull request by wischli**
# Description

* Adds `ForeignInvestmentBenchmarkHelper` which provides the setup required for benchmarking the heaviest foreign investment call
* Adds `benchmarking` mod to `pallet-liquidity-pools` which benches an incoming `CollectRedeem` message by executing `handle_collect_redemption`
  * NOTE: Skipped the `InboundQueue` since this complicates the investor setup: Right now, the `investor` account is created during the pool creation of `pool_system::PoolBenchmarkHelper` and thus just some Substrate account id. However, for the `InboundQueue` it would have to be a `AccountConverter<Runtime, LocationToAccountId>::convert((sender_domain, investor_bytes))`.
* Feature creep: Renames `pallet_foreign_investments::<T>::Config::Rate` to `BalanceRatio` 

## TODO

- [ ] Discuss how to move forward: Do we want to defensively stick to the hardcoded weights?
  * If so, we should not add `LiquidityPools` to the runtimes' benchmark list 
  * Else, I would continue writing benchmarks for existing extrinsics such that we have everything covered properly: an inbound messages by the FI worst case (`handle_collect_redemption`) and extrinsics by their corresponding benchmarks
- [ ] Maybe: Right now, the `pallet-liquidity-pools` benchmarks cannot be unit tested, because the existing mock is insufficient. If running unit tests is desired, I propose to enable this in a subsequent PR.

## Justification about heaviest FI call

* In the context of `pallet-foreign-investments`, the heaviest action (measured by amount of reads and writes) is collecting a redemption when there exists an `InvestmentState` which swaps from `foreign` to `pool`
* By collecting the redemption, a swap from `pool` to `foreign` emerges which needs to be resolved in `handle_concurrent_swap_orders` by comparing with the existing swap from the investment

## About the weight

For presentation purposes, I temporarily added the generated weight file. This must be removed before merging as otherwise executing benchmarks for `pallet-liquidity-pools` would remove all hardcoded values.

When looking at the results of executing the benchmark on my local machine, the total weight is ~55% of the current hardcoded one. The PoV size is ~80% of the hardcoded one. For details, see the [WIP Weights Sheet](https://docs.google.com/spreadsheets/d/1X0N-TJg1R3Lsr1ILqW0iVWdl5etlXYXPMobzFWpxdCw/edit#gid=0). I suppose the locally generated weights are close enough to our benchmark runner, since the results are similar.

```
		Weight::from_parts(236_000_000, 71940)
			.saturating_add(T::DbWeight::get().reads(20))
			.saturating_add(T::DbWeight::get().writes(17))
```

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [ ] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: pallets/foreign-investments/src/lib.rs
Comment (wischli) [inline pallets/foreign-investments/src/lib.rs:155]: Sorry for feature creeping. This change does not have any functional impact.

File: pallets/liquidity-pools/src/weights_generated.rs
Comment (wischli) [inline pallets/liquidity-pools/src/weights_generated.rs:2]: NOTE: Committed solely for demonstration purpose, will be removed before merging!

File: pallets/foreign-investments/src/impls/benchmark_utils.rs
Comment (lemunozm) [inline pallets/foreign-investments/src/impls/benchmark_utils.rs:73]: Comment: I really like how the `bench_*` methods are called inside `bench_*` methods, leaving a clean API to use from the benchmark side with 0 knowledge about what's happening under the hood. i.e. the `pool_id` used is totally transparent here ❤️

File: pallets/pool-system/src/impls.rs
Comment (lemunozm) [inline pallets/pool-system/src/impls.rs:531]: Just curious, why is this minting now needed?

**Review by lemunozm — COMMENTED** [review]
Love how clean are you getting this given that is quite difficult benchmark. 💯

**Review by mustermeiszer — DISMISSED** [review]
Sick! Whats your final feeling regarding the adaption of the `process_msg` weight? 

I am personally leaning torwards keeping it as is an heavier. But might be paranoia... 

@wischli could you include the changes from #1562, then we do not need to wait for CI 2 times.

File: libs/traits/src/benchmarking.rs
Comment (wischli) [reply libs/traits/src/benchmarking.rs]: [239fa89](https://github.com/centrifuge/centrifuge-chain/pull/1565/commits/239fa899c9f38c61e0dd0003f0fb5a38aa01d46e)

File: libs/traits/src/benchmarking.rs
Comment (lemunozm) [inline libs/traits/src/benchmarking.rs:70]: I think it doesn't need to be generalized here. The new struct can just be used directly and even can be in the traits module (i.e. just before the trait). I mean, it's not a type the user needs to configure in runtime or similar. It's just an aggregation of fields to avoid using a nameless tuple. WDYT?

Comment (wischli) [reply libs/traits/src/benchmarking.rs:70]: That was my initial thought but then I assumed you'd prefer it to be generic. Will follow your proposal then to make it absolutely explicit!

File: pallets/pool-system/src/impls.rs
Comment (wischli) [reply pallets/pool-system/src/impls.rs:531]: This was taken from the integration tests when I was debugging `BalanceTooLow` errors. So it might not be necessary but I believe it still is.

When we process a redemption of tranche tokens, the pool account needs to have native balance which is exchanged for these tranche tokens. 

In contrast, we don't need to fund the pool account with tranche tokens as these are minted when processing an investment.

**Review by mustermeiszer — APPROVED** [review]

**Pull request merged**
